src/zfm20/zfm20.cxx

```diff
@@ -33,154 +33,53 @@ using namespace std;
-                                  ": mraa_uart_init() failed");
-      return;
-    }
-
-  // This requires a recent MRAA (1/2015)
-  const char *devPath = mraa_uart_get_dev_path(m_uart);
-
-  if (!devPath)
-    {
-      throw std::runtime_error(std::string(__FUNCTION__) +
-                               ": mraa_uart_get_dev_path() failed");
-      return;
-    }
-
-  // now open the tty
-  if ( (m_ttyFd = open(devPath, O_RDWR)) == -1)
-    {
-      throw std::runtime_error(std::string(__FUNCTION__) +
-                               ": open of " + 
-                               string(devPath) + " failed: " +
-                               string(strerror(errno)));
-      return;
-    }
-
   // Set the default password and address
   setPassword(ZFM20_DEFAULT_PASSWORD);
   setAddress(ZFM20_DEFAULT_ADDRESS);
 
   initClock();
 }
 
-ZFM20::~ZFM20()
-{
-  if (m_ttyFd != -1)
-    close(m_ttyFd);
-
-  mraa_deinit();
-}
-
-bool ZFM20::dataAvailable(unsigned int millis)
-{
-  if (m_ttyFd == -1)
-    return false;
-
-  struct timeval timeout;
-
-  // no waiting
-  timeout.tv_sec = 0;
-  timeout.tv_usec = millis * 1000;
-
-  fd_set readfds;
-
-  FD_ZERO(&readfds);
-
-  FD_SET(m_ttyFd, &readfds);
-  
-  if (select(m_ttyFd + 1, &readfds, NULL, NULL, &timeout) > 0)
-    return true;                // data is ready
-  else
-    return false;
-}
-
 int ZFM20::readData(char *buffer, int len)
 {
-  if (m_ttyFd == -1)
-    return(-1);
-
-  if (!dataAvailable(defaultDelay))
+  if (!m_uart.dataAvailable(defaultDelay))
     return 0;               // timed out
 
-  int rv = read(m_ttyFd, buffer, len);
+  int rv = m_uart.read(buffer, len);
 
   if (rv < 0)
-    {
-      throw std::runtime_error(std::string(__FUNCTION__) +
-                               ": read() failed: " +
-                               string(strerror(errno)));
-      return rv;
-    }
+    throw std::runtime_error(std::string(__FUNCTION__) +
+            ": Uart::read() failed: " + string(strerror(errno)));
 
   return rv;
 }
 
 int ZFM20::writeData(char *buffer, int len)
 {
-  if (m_ttyFd == -1)
-    return(-1);
-
   // first, flush any pending but unread input
-  tcflush(m_ttyFd, TCIFLUSH);
+  m_uart.flush();
 
-  int rv = write(m_ttyFd, buffer, len);
+  int rv = m_uart.write(buffer, len);
 
   if (rv < 0)
-    {
       throw std::runtime_error(std::string(__FUNCTION__) +
-                               ": write() failed: " +
+                               ": Uart::write() failed: " +
                                string(strerror(errno)));
-      return rv;
-    }
 
   if (rv == 0)
-    {
       throw std::runtime_error(std::string(__FUNCTION__) +
-                               ": write() failed, no bytes written");
-      return rv;
-    }
+                               ": Uart::write() failed, no bytes written");
 
-  tcdrain(m_ttyFd);
+  m_uart.flush();
 
   return rv;
 }
 
-bool ZFM20::setupTty(speed_t baud)
+bool ZFM20::setupTty(uint32_t baud)
 {
```

Comment:
If you are amenable to API changes, I would just remove this function and have the baudrate, passed as a standard int, to the ctor, and then set it there.  Or, this function could just be renamed to setBaudrate() instead (but still accepting a standard int rather than int32_t).  As you know by now, almost all of what setupTty() did is handled by MRAA now anyway.